### PR TITLE
soc: arm: stm32: Updated SRAM Region Definitions

### PR DESCRIPTION
**Pull request by Hobbs1210**
Support for STM32F7XX MCU's that have 512k of SRAM.

### CODE REVIEW

**Review by ioannisg — CHANGES_REQUESTED** [review]

**Review by ioannisg — COMMENTED** [review]
Commit title prefix should, probably be: "soc: arm: stm32:"
It's not really an ARM architecture patch

**Comment by ioannisg** [comment]
adding this to 1.15, unless it is a bug affecting current boards/SOCs. If so, please indicate and re-scope.

File: soc/arm/st_stm32/common/arm_mpu_mem_cfg.h [hunk 1]
```diff
@@ -77,6 +77,10 @@
 #define REGION_SRAM_0_SIZE REGION_256K
 #define REGION_SRAM_1_START 0x40000
 #define REGION_SRAM_1_SIZE REGION_128K
+#elif CONFIG_SRAM_SIZE == 512
```
[the remark this answers is not in the document]
Comment (Hobbs1210) [reply soc/arm/st_stm32/common/arm_mpu_mem_cfg.h]: Good catch, fixed!

**Review by ioannisg — COMMENTED** [review]
So, the patch is now OK to me, however, the commit history looks confusing.
Could you please clean this up? I.e. commit this diff in a single commit, and force-push to this pull-request, so the PR is a single commit. Refer to the Zephyr contribution guidelines if needed :)

Then we can approve the PR and add it to the queue for merging to master.

**Review by erwango — CHANGES_REQUESTED** [review]
As requested by @ioannisg, please restructure your change into a single commit and fix commit message issues reported by @zephyrbot

**Comment by Hobbs1210** [comment]
Fixed per request

**Comment by psidhu** [comment]
@erwango might know better than me, but shouldn't you split this up into two REGION_256K's?

For example, even the 128k support uses two REGION_64K's to split the RAM up instead of a single REGION_128K.

**Comment by ioannisg** [comment]
> @erwango might know better than me, but shouldn't you split this up into two REGION_256K's?
> 
> For example, even the 128k support uses two REGION_64K's to split the RAM up instead of a single REGION_128K.

Yes, it looks like this file needs rework.
So, the rule should be the following:
SRAM sizes that are power-of-two shall use a single SRAM MPU region.

**Comment by Hobbs1210** [comment]
> > @erwango might know better than me, but shouldn't you split this up into two REGION_256K's?
> > For example, even the 128k support uses two REGION_64K's to split the RAM up instead of a single REGION_128K.
> 
> Yes, it looks like this file needs rework.
> So, the rule should be the following:
> SRAM sizes that are power-of-two shall use a single SRAM MPU region (Region 0).
> If not a power-of-two, we shall use as few MPU regions as possible.
> @Hobbs1210 could you please, enhance this PR so it fixes all issues?

@ioannisg please review the latest push. please let me know if there are any further changes required.

**Comment by erwango** [comment]
@Hobbs1210, you'll need to also fix soc/arm/st_stm32/common/arm_mpu_regions.c which was implemented on the assumption there would always be 2 RAM zones.
 
Btw, this was implemented this way to have same number of available zones, whether RAM size is a power of 2 or not. This indeed can be optimized and save one MPU region when size is directly a power of 2. Which leaves one region free in this case. 
@ioannisg, is there anything we can do with this available MPU region?

**Comment by Hobbs1210** [comment]
> @Hobbs1210, you'll need to also fix soc/arm/st_stm32/common/arm_mpu_regions.c which was implemented on the assumption there would always be 2 RAM zones.
> 
> Btw, this was implemented this way to have same number of available zones, whether RAM size is a power of 2 or not. This indeed can be optimized and save one MPU region when size is directly a power of 2. Which leaves one region free in this case.
> @ioannisg, is there anything we can do with this available MPU region?

@erwango please review my latest push.

**Review by erwango — CHANGES_REQUESTED** [review]
Please add a one liner comment in commit message to get git check happy

**Comment by ioannisg** [comment]
> @ioannisg, is there anything we can do with this available MPU region?

Yes, we can use it for application memory partitions. The more we have the better.

**Comment by erwango** [comment]
> Yes, we can use it for application memory partitions. The more we have the better.

What bothers me is you'll have one extra partition depending on the memory size, so a 256K config will enjoy one more partition than a 384K..

**Review by erwango — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]
thanks

**Pull request merged**